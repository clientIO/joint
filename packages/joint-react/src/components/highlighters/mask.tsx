import type { FC, PropsWithChildren } from 'react';
import { forwardRef, useCallback, useMemo } from 'react';
import type { OnCreateHighlighter } from './custom';
import { Custom } from './custom';
import type { dia } from '@joint/core';
import { highlighters } from '@joint/core';

export interface MaskHighlighterProps extends React.SVGProps<SVGPathElement>, PropsWithChildren {
  /**
   * The layer on which the mask will be rendered.
   */
  readonly layer?: string;
  /**
   * A CSS selector string for targeting elements.
   */
  readonly selector?: string;
  /**
   * Child elements to render inside the mask.
   */
  readonly children?: React.ReactNode | null | false;
  /**
   * The space between the stroke and the element
   */
  readonly padding?: number;
  /**
   * If the highlighter is disabled or not.
   */
  readonly isHidden?: boolean;
}

const DEFAULT_MASK_HIGHLIGHTER_PROPS: MaskHighlighterProps = {
  stroke: '#4666E5',
  strokeWidth: 3,
  strokeLinejoin: 'round',
  fill: 'none',
};

// eslint-disable-next-line jsdoc/require-jsdoc
function Component(props: MaskHighlighterProps, forwardedRef: React.Ref<SVGElement>) {
  const { layer, children, padding, isHidden, ...svgAttributes } = props;
  const options = useMemo((): dia.HighlighterView.Options => {
    const data: dia.HighlighterView.Options = {
      layer,
      attrs: {
        ...DEFAULT_MASK_HIGHLIGHTER_PROPS,
        ...svgAttributes,
      },
    };
    if (padding !== undefined) {
      data.padding = padding;
    }
    return data;
  }, [layer, padding, svgAttributes]);

  const onAdd: OnCreateHighlighter = useCallback((cellView, element, highlighterId, hOptions) => {
    return highlighters.mask.add(cellView, element, highlighterId, hOptions);
  }, []);

  return (
    <Custom options={options} ref={forwardedRef} onCreateHighlighter={onAdd} isHidden={isHidden}>
      {children}
    </Custom>
  );
}

/**
 * Mask highlighter component.
 * Adds a stroke around an arbitrary cell view's SVG node.
 * @see https://docs.jointjs.com/api/highlighters/#mask
 * @group Components
 * @example
 * ```tsx
 * import { Highlighter } from '@joint/react'
 * return <Highlighter.Mask />
 * ```
 */
export const Mask: FC<MaskHighlighterProps> = forwardRef(Component);
