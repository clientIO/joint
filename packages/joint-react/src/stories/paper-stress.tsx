/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable react-perf/jsx-no-new-object-as-prop */
/* eslint-disable sonarjs/pseudo-random */
/* eslint-disable no-console */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */
import { dia, shapes } from '@joint/core'
import { ReactElement } from '../models/react-element'
import { useElements } from '../hooks/use-elements'
import { useSetCells } from '../hooks/use-set-cells'
import { useGraph } from '../hooks/use-graph'
import type { PaperStory } from './paper.stories'
import { GraphProvider } from '../components/graph-provider'
import { PaperProvider } from '../components/paper-provider'
import type { RenderElement } from '../components/paper'
import { Paper } from '../components/paper'
import { useCallback, useRef } from 'react'
import { HtmlElement } from '../components/html-element'

const paperStoryOptions: dia.Paper.Options = {
  width: 400,
  height: 400,
  background: { color: '#f8f9fa' },
  gridSize: 2,
}

function createElements(xCount: number, yCount: number) {
  const elements = []
  const ELEMENT_SIZE = 50
  const MARGIN = 2
  for (let x = 0; x < xCount; x++) {
    for (let y = 0; y < yCount; y++) {
      elements.push(
        new ReactElement({
          id: `${x}-${y}`,
          position: { x: x * (ELEMENT_SIZE + MARGIN), y: y * (ELEMENT_SIZE + MARGIN) },
          size: { width: ELEMENT_SIZE, height: ELEMENT_SIZE },
          attrs: {
            label: { text: `${x}-${y}` },
          },
        })
      )
      // add a link to the previous element
      if (x > 0) {
        elements.push(
          new shapes.standard.Link({
            id: `${x - 1}-${y}-${x}-${y}`,
            source: { id: `${x - 1}-${y}` },
            target: { id: `${x}-${y}` },
          })
        )
        // add link to next element
        if (y > 0) {
          elements.push(
            new shapes.standard.Link({
              id: `${x}-${y - 1}-${x}-${y}`,
              source: { id: `${x}-${y - 1}` },
              target: { id: `${x}-${y}` },
            })
          )
        }
      }
    }
  }
  return elements
}

const WIDTH_ITEMS = 15
const HEIGHT_ITEMS = 30

function RandomChange() {
  const elementsSize = useElements((items) => items.length)
  const setCells = useSetCells()
  const graph = useGraph()
  return (
    <>
      <button
        onClick={() => {
          console.time('Random move')
          setCells((previousCells) =>
            previousCells.map((cell) => {
              if (cell instanceof ReactElement) {
                cell.set({
                  position: { x: Math.random() * 1000, y: Math.random() * 1000 },
                })
                return cell
              }
              return cell
            })
          )
          console.timeEnd('Random move')
        }}
      >
        Random move {elementsSize} elements - set With hook
      </button>
      <button
        onClick={() => {
          graph.startBatch('Random move')
          const elements = graph.getElements()
          console.time('Random move')
          for (const element of elements) {
            if (element instanceof ReactElement) {
              element.set({
                position: { x: Math.random() * 1000, y: Math.random() * 1000 },
              })
            }
          }
          console.timeEnd('Random move')
          graph.stopBatch('Random move')
        }}
      >
        Random move {elementsSize} elements - set With Graph
      </button>
    </>
  )
}

function createGraph() {
  const graph = new dia.Graph({}, { cellNamespace: { ...shapes, ReactElement } })
  graph.addCells(createElements(WIDTH_ITEMS, HEIGHT_ITEMS))
  return graph
}
export const PaperStressTestNative: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    const graph = useRef(createGraph()).current
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider graph={graph}>
        <RandomChange />
        <div style={{ display: 'flex', flex: 1 }}>
          <PaperProvider {...paperStoryOptions}>
            <Paper />
          </PaperProvider>
        </div>
      </GraphProvider>
    )
  },
}

export const PaperStressTestReact: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    const graph = useRef(createGraph()).current

    console.log('re-render WithHooksAPI')

    const renderElement: RenderElement = useCallback((element) => {
      return (
        <HtmlElement
          width={element.width}
          height={element.height}
          style={{
            fontSize: 12,
            background: 'blue',
            width: '100%',
            height: '100%',
            overflow: 'hidden',
          }}
          onClick={() => console.log('Click from React')}
        >
          {JSON.stringify(element.x)}
        </HtmlElement>
      )
    }, [])
    return (
      <GraphProvider graph={graph}>
        <RandomChange />
        <div style={{ display: 'flex', flex: 1 }}>
          <Paper {...paperStoryOptions} renderElement={renderElement} />
        </div>
      </GraphProvider>
    )
  },
}
