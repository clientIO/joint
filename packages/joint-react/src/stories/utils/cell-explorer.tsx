/**
 * This is used only for storybook stories. Internally.
 */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */
/* eslint-disable react-perf/jsx-no-new-object-as-prop */
import type { dia } from '@joint/core';

interface CellExplorerProps {
  readonly cell: dia.Cell.JSON;
  readonly onChange?: (cell: dia.Cell.JSON) => void;
}

interface Props {
  readonly keyName: string;
  readonly parentKey?: string;
  readonly value: unknown;
  readonly onChange: (newValue: unknown) => void;
}
const MARGIN = '8px';
function EditableField({ keyName, parentKey, value, onChange }: Props) {
  const handleChange = (key: string, newValue: unknown) => {
    if (typeof value === 'object' && value !== null) {
      onChange({ ...value, [key]: newValue });
    } else {
      onChange(newValue);
    }
  };

  const parseValue = (inputValue: string) => {
    if (typeof value === 'number') {
      return Number.isNaN(Number(inputValue)) ? value : Number(inputValue);
    }
    return inputValue;
  };

  if (typeof value === 'object' && value !== null) {
    return (
      <div style={{ marginBottom: MARGIN }}>
        <div>
          {Object.entries(value).map(([key, value_]) => (
            <EditableField
              key={key}
              keyName={key}
              parentKey={parentKey ? `${parentKey}.${keyName}` : keyName}
              value={value_}
              onChange={(newValue) => handleChange(key, newValue)}
            />
          ))}
        </div>
      </div>
    );
  }

  return (
    <div style={{ marginBottom: MARGIN }}>
      <label style={{ fontWeight: 'bold', marginRight: MARGIN }}>
        {parentKey ? `${parentKey}.${keyName}` : keyName}:{' '}
      </label>
      <input
        type="text"
        value={String(value)}
        onChange={(event) => onChange(parseValue(event.target.value))}
        style={{ marginLeft: MARGIN }}
      />
    </div>
  );
}

function CellExplorer({ cell, onChange }: CellExplorerProps) {
  const handleInputChange = (key: string, value: unknown) => {
    if (onChange) {
      onChange({ ...cell, [key]: value });
    }
  };

  return (
    <div style={{ border: '1px solid #ddd', padding: MARGIN, margin: '4px' }}>
      <h4>Cell ID: {cell.id}</h4>
      {Object.entries(cell).map(([key, value]) => {
        return (
          <EditableField
            key={key}
            keyName={key}
            value={value}
            onChange={(newValue) => handleInputChange(key, newValue)}
          />
        );
      })}
    </div>
  );
}

interface CellsExplorerProps {
  readonly elements: dia.Cell.JSON[];
  readonly onChange: (cells: dia.Cell.JSON[]) => void;
}

function CellsExplorer({ elements: cells, onChange }: CellsExplorerProps) {
  return (
    <div style={{ display: 'flex', flexDirection: 'row', flexWrap: 'wrap' }}>
      {cells.map((cell) => {
        if (!cell) {
          return null;
        }
        return (
          <CellExplorer
            key={cell.id}
            cell={cell}
            onChange={(newCell) => {
              const updatedCells = cells.map((c) => {
                if (!c) {
                  return c;
                }
                if (c.id === newCell.id) {
                  return newCell;
                }
                return c;
              });
              onChange(updatedCells);
            }}
          />
        );
      })}
    </div>
  );
}

export { CellExplorer, CellsExplorer };
