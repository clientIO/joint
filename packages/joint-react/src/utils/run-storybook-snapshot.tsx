/* eslint-disable unicorn/consistent-function-scoping */
/* eslint-disable jsdoc/require-jsdoc */
import { render } from '@testing-library/react';
import { GraphProvider, Paper } from '../components';

interface Options<StorybookOptions> {
  Component: React.ComponentType<StorybookOptions>;
  name: string;
  // we have to use any here because we don't know the type of the story
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  stories: Record<string, any>;
  withRenderElementWrapper?: boolean;
}

function normalizeSnapshot(html: string) {
  const cleaned = html.replaceAll(/id="v-\d+"/g, 'id="v-*"');

  return cleaned;
}

/**
 * Runs a snapshot test for each story in the provided options.
 * @param options - The options for the snapshot test.
 * @param options.Component - The component to render.
 * @param options.name - The name of the test suite.
 * @param options.stories - The stories to test.
 */
export function runStorybookSnapshot<StorybookOptions>(options: Options<StorybookOptions>) {
  const { stories, Component, name, withRenderElementWrapper } = options;
  const keys = Object.keys(stories).filter((key) => key !== 'default');
  describe(name, () => {
    it.each(keys)('%p', (key) => {
      const story = stories[key as keyof typeof stories];

      const props = ('args' in story ? story.args : {}) as StorybookOptions & {
        children?: React.ReactNode;
      };

      let wrapper = function ({ children }: { children: React.ReactNode }) {
        return <>{children}</>;
      };
      if (withRenderElementWrapper) {
        wrapper = function ({ children }: { children: React.ReactNode }) {
          return (
            <GraphProvider>
              <Paper renderElement={children as never} />
            </GraphProvider>
          );
        };
      }
      // ACT
      const { container } = render(<Component {...props} />, { wrapper });

      // Clean volatile parts before asserting
      const cleanedHTML = normalizeSnapshot(container.innerHTML);
      expect(cleanedHTML).toMatchSnapshot(`${name}-${key}`);
    });
  });
}
