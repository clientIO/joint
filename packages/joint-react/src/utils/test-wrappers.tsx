import { useCallback } from 'react';
import { GraphProvider, Paper, type GraphProps, type PaperProps } from '../components';

/**
 * This wrapper is used to render a graph provider.
 * It is used in the tests to render a graph provider.
 * @param props - The props for the graph provider.
 * @returns - The wrapper.
 * @internal
 * @group utils
 * @description
 * This wrapper is used to render a graph provider.
 */
export function graphProviderWrapper(props: GraphProps): React.JSXElementConstructor<{
  children: React.ReactNode;
}> {
  return function GraphProviderWrapper({ children }) {
    return <GraphProvider {...props}>{children}</GraphProvider>;
  };
}

interface Options {
  paperProps?: PaperProps;
  graphProps?: GraphProps;
}
/**
 * This wrapper is used to render a paper with a graph provider.
 * It is used in the tests to render a paper with a graph provider.
 * @param options - The options for the wrapper.
 * @param options.paperProps - The props for the paper.
 * @param options.graphProps - The props for the graph provider.
 * @returns - The wrapper.
 * @internal
 * @group utils
 */
export function paperRenderElementWrapper(options: Options): React.JSXElementConstructor<{
  children: React.ReactNode;
}> {
  const { paperProps, graphProps } = options;
  return function GraphProviderWrapper({ children }) {
    const renderElement = useCallback(() => {
      return children;
    }, [children]);
    return (
      <GraphProvider {...graphProps}>
        <Paper {...paperProps} renderElement={renderElement}></Paper>
      </GraphProvider>
    );
  };
}

export const simpleRenderElementWrapper = paperRenderElementWrapper({
  graphProps: {
    initialElements: [
      {
        id: '1',
        width: 97,
        height: 99,
      },
    ],
  },
});
