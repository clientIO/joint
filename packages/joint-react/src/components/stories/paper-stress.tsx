/* eslint-disable react-perf/jsx-no-new-object-as-prop */
/* eslint-disable sonarjs/pseudo-random */
/* eslint-disable no-console */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */
import { dia, shapes } from '@joint/core'
import { ReactElement } from '../../models/react-element'
import { GraphProvider } from '../graph-provider'
import { Paper } from '../paper'
import { PaperProvider } from '../paper-provider'
import type { PaperStory } from './paper.stories'
import { useSetCells } from '../../hooks/use-set-cells'
import { updateGraph } from '../../utils/update-graph'
import { useElements } from '../../hooks/use-elements'

const paperStoryOptions: dia.Paper.Options = {
  width: 400,
  height: 400,
  background: { color: '#f8f9fa' },
  gridSize: 2,
}
const graph = new dia.Graph({}, { cellNamespace: { ...shapes, ReactElement } })
const notUsedGraph = new dia.Graph({}, { cellNamespace: { ...shapes, ReactElement } })
function createElements(xCount: number, yCount: number) {
  const elements = []
  const ELEMENT_SIZE = 50
  const MARGIN = 2
  for (let x = 0; x < xCount; x++) {
    for (let y = 0; y < yCount; y++) {
      elements.push(
        new ReactElement({
          id: `${x}-${y}`,
          position: { x: x * (ELEMENT_SIZE + MARGIN), y: y * (ELEMENT_SIZE + MARGIN) },
          size: { width: ELEMENT_SIZE, height: ELEMENT_SIZE },
          attrs: {
            label: { text: `${x}-${y}` },
            body: { fill: 'blue', stroke: 'black' },
          },
        })
      )
      // add a link to the previous element
      if (x > 0) {
        elements.push(
          new shapes.standard.Link({
            id: `${x - 1}-${y}-${x}-${y}`,
            source: { id: `${x - 1}-${y}` },
            target: { id: `${x}-${y}` },
          })
        )
      }
    }
  }
  return elements
}

graph.addCells(createElements(15, 30))
notUsedGraph.addCells(createElements(15, 30))
function RandomChange() {
  const elementsSize = useElements((items) => items.map((item) => item.id))
  const setCells = useSetCells()
  console.log('re-render RandomChange Component', elementsSize)
  return (
    <>
      <button
        onClick={() => {
          console.time('Random move')
          setCells((previousCells) =>
            previousCells.map((cell) => {
              if (cell instanceof ReactElement) {
                cell.set({
                  position: { x: Math.random() * 1000, y: Math.random() * 1000 },
                })
                return cell
              }
              return cell
            })
          )
          console.timeEnd('Random move')
        }}
      >
        {/* Random move {elementsSize.length} elements */}
      </button>

      <button
        onClick={() => {
          console.time('Random move with notUsedGraph')
          const oldCells = notUsedGraph.getCells()
          updateGraph(
            notUsedGraph,
            oldCells.map((cell) => {
              if (cell instanceof ReactElement) {
                cell.set({
                  position: { x: Math.random() * 1000, y: Math.random() * 1000 },
                })
                return cell
              }
              return cell
            })
          )
          console.timeEnd('Random move with notUsedGraph')
        }}
      >
        Random move with notUsedGraph - only prints console.logs
      </button>
    </>
  )
}

export const PaperStressTestNative: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider graph={graph}>
        <RandomChange />
        <div style={{ display: 'flex', flex: 1 }}>
          <PaperProvider {...paperStoryOptions}>
            <Paper />
          </PaperProvider>
        </div>
      </GraphProvider>
    )
  },
}

export const PaperStressTestReact: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider graph={graph}>
        <RandomChange />
        <div style={{ display: 'flex', flex: 1 }}>
          {/* <PaperProvider {...paperStoryOptions}> */}
          <Paper
            {...paperStoryOptions}
            // elementSelector={(cell) => {
            //   switch (cell instanceof ReactElement) {
            //     case true: {
            //       return { id: cell.id, xPosition: cell.attributes.position?.x }
            //     }
            //     default: {
            //       return {
            //         id: cell.id,
            //       }
            //     }
            //   }
            // }}
            renderElement={(element) => {
              console.log('re-render renderElement')
              return (
                <div style={{ fontSize: 12 }} onClick={() => console.log('Click from React')}>
                  {JSON.stringify(element.attributes.position)}
                </div>
              )
            }}
          />
          {/* </PaperProvider> */}
        </div>
      </GraphProvider>
    )
  },
}
