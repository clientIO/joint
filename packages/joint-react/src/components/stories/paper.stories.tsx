/* eslint-disable sonarjs/no-small-switch */

/* eslint-disable no-console */
/* eslint-disable react-perf/jsx-no-new-object-as-prop */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */

import type { Meta, StoryObj } from '@storybook/react'
import type { dia } from '@joint/core'

import { shapes } from '@joint/core'
import { GraphProvider } from '../graph-provider'
import { PaperProvider } from '../paper-provider'
import { Paper } from '../paper'
import { useGraphStore } from '../../hooks/use-graph-store'
import { useGraphCells } from '../../hooks/use-graph-cells'
import { CellsExplorer } from './cell-explorer'
import { ReactElement } from '../../models/react-element'

import { useSetGraphCells } from '../../hooks/use-set-graph-cells'

const paperStoryOptions: dia.Paper.Options = {
  width: 400,
  height: 400,
  background: { color: '#f8f9fa' },
  gridSize: 2,
}

// elements to add to the graph
const ELEMENTS = () => {
  return [
    new shapes.standard.Rectangle({
      id: '1',
      position: { x: 100, y: 30 },
      size: { width: 100, height: 40 },
      attrs: {
        label: { text: 'rect-1' },
        body: { fill: 'blue' },
      },
    }),
    new shapes.standard.Rectangle({
      id: '2',
      position: { x: 100, y: 100 },
      size: { width: 100, height: 40 },
      attrs: {
        label: { text: 'rect-2' },
        body: { fill: 'red' },
      },
    }),
    new shapes.standard.Link({
      source: { id: '1' },
      target: { id: '2' },
      attrs: {
        line: { stroke: 'blue', targetMarker: { name: 'classic', size: 8 } },
      },
    }),
    new ReactElement({
      id: '3',
      position: { x: 100, y: 200 },
      size: { width: 100, height: 40 },
      attrs: {
        body: { fill: 'green' },
      },
    }),

    new ReactElement({
      id: '4',
      position: { x: 100, y: 300 },
      size: { width: 100, height: 40 },
      attrs: {
        body: { fill: 'green' },
      },
    }),
  ]
}
function UpdateCellsViaGraphApi() {
  const { graph } = useGraphStore()

  return (
    <div>
      <div>Updating cells via graph API</div>
      <button onClick={() => graph.clear()}>Remove all cells</button>
      <button onClick={() => graph.addCells(ELEMENTS())}>Add all cells</button>
    </div>
  )
}

const meta: Meta<typeof Paper> = {
  title: 'Components/Paper',
  component: Paper,
}

export default meta
export type PaperStory = StoryObj<typeof Paper>

function CellsExplorerViaHook() {
  const cells = useGraphCells((cell) => cell.toJSON())
  const setCells = useSetGraphCells()
  return <CellsExplorer cells={cells} onChange={setCells} />
}

export const WithHooksAPI: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider>
        <UpdateCellsViaGraphApi />
        <div style={{ display: 'flex', flex: 1 }}>
          <PaperProvider {...paperStoryOptions}>
            <Paper
              elementSelector={(cell) => {
                switch (cell instanceof ReactElement) {
                  case true: {
                    return { id: cell.id, x: cell.attributes.position?.x }
                  }
                  default: {
                    return {
                      id: cell.id,
                    }
                  }
                }
              }}
              renderElement={(cell) => (
                <div style={{ width: '100%', height: '10%' }} onClick={() => console.log('CLICK')}>
                  x: {cell.x}
                </div>
              )}
            />
          </PaperProvider>
          <CellsExplorerViaHook />
        </div>
      </GraphProvider>
    )
  },
}

export { PaperStressTestNative, PaperStressTestReact } from './paper-stress'
