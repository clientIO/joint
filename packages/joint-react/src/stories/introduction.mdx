import { Meta } from '@storybook/blocks';
import { getAPIDocumentationLink, getAPIPropsLink } from './utils/get-api-documentation-link';

<Meta title="Introduction" />

# Introduction to @joint/react

**@joint/react** is a React wrapper for [JointJS](https://www.jointjs.com/), designed to make creating and managing interactive diagrams and graphs in React applications simple and intuitive. It provides React components, hooks, and utilities to efficiently manage nodes, links, and user interactions.

## üìå Core Components

### 1. {getAPIDocumentationLink('GraphProvider')}
The `GraphProvider` component manages a shared [JointJS Graph instance](https://docs.jointjs.com/api/dia/Graph/) to handle the state of your diagram. Wrap it around any components that interact with the graph.

```tsx
import { GraphProvider } from '@joint/react';

<GraphProvider>
  {/* Components like Paper for rendering nodes and edges */}
</GraphProvider>
```

### 2. {getAPIDocumentationLink('Paper', 'variables')}
The `Paper` component wraps [JointJS Paper](https://docs.jointjs.com/learn/quickstart/paper/) to render nodes and links. Use the {getAPIPropsLink('Paper', 'renderElement')} prop to define how nodes are displayed.

```tsx
import { Paper } from '@joint/react';

const renderElement = (element) => (
  <rect width={element.size().width} height={element.size().height} fill="cyan" />
);

<Paper width={800} height={600} renderElement={renderElement} />
```

### 3. **Embedding HTML in SVG Nodes**
While JointJS primarily uses SVG, you can embed HTML content inside nodes using SVG's `<foreignObject>`:

```tsx
const renderElement = ({ width, height }) => (
  <foreignObject width={width} height={height}>
    <div style={{ background: 'lightgray' }}>
      HTML Content here
    </div>
  </foreignObject>
);
```

## üõ†Ô∏è Core Hooks and Utilities

### üîπ Accessing Elements
- {getAPIDocumentationLink('useElements')}: Retrieve all diagram elements (requires `GraphProvider` context).
- {getAPIDocumentationLink('useElement')}: Retrieve individual element data, typically used within `renderElement`.

### üîπ Modifying Elements
- {getAPIDocumentationLink('useUpdateElement')}: Update existing elements in the diagram.

### üîπ Graph and Paper Instances
- {getAPIDocumentationLink('useGraph')}: Access the JointJS [Graph instance](https://docs.jointjs.com/api/dia/Graph/) directly.
- {getAPIDocumentationLink('usePaper')}: Access the JointJS [Paper instance](https://docs.jointjs.com/learn/quickstart/paper/) directly.

### üîπ Creating Nodes and Links
- {getAPIDocumentationLink('createElements')}: Utility for creating nodes.

```ts
import { createElements } from '@joint/react';

const initialElements = createElements([
  { id: '1', type: 'rect', x: 10, y: 10, width: 100, height: 100 },
]);
```

- {getAPIDocumentationLink('createLinks')}: Utility for creating links between nodes.

```ts
import { createLinks } from '@joint/react';

const initialLinks = createLinks([
  { source: '1', target: '2', id: '1-2' },
]);
```

---

## How It Works
Under the hood, **@joint/react** listens to changes in the `dia.Graph`, which acts as the single source of truth. When you update the graph‚Äîsuch as adding or modifying cells‚Äîthe React components automatically observe and react to these changes, keeping the UI in sync.

Hooks like `useUpdateElement` provide a convenient way to update the graph, but you can also directly access the graph using `useGraph()` and call methods like `graph.setCells()`.

---

## Known Issues

### Avoid Certain CSS Properties in `<foreignObject>`
Some CSS properties can cause rendering issues in Safari when used inside an SVG `<foreignObject>`. To ensure compatibility, avoid the following properties:

- `position` (other than `static`)
- `-webkit-transform-style`
- `-webkit-backface-visibility`
- `transition`
- `transform`

### Recommended Workaround
If you need to use HTML inside an SVG with cross-browser support:

- Use minimal CSS inside `<foreignObject>`.
- Stick to static positioning and avoid CSS transforms.
- Consider overlaying HTML outside the SVG using absolute positioning.

### Flickering
React's asynchronous rendering can cause flickering when dynamically adding ports or resizing elements. We are aware of this issue and are working on a fix.

### Controlled Mode
Currently, **@joint/react** uses `useSyncExternalStore` to listen to graph changes. The graph is the source of truth, so `initialElements` and `initialLinks` are only used during initialization. To modify the state, update the graph directly using hooks like `useGraph`, `useUpdateElement`, or `useCreateElement`. A fully controlled mode is under development.
