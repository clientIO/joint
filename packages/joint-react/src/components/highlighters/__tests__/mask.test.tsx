import { render, waitFor } from '@testing-library/react';
import { runStorybookSnapshot } from '../../../utils/run-storybook-snapshot';
import { Mask } from '../mask';
import * as stories from '../mask.stories';
import { simpleRenderElementWrapper } from '../../../utils/test-wrappers';
runStorybookSnapshot({
  Component: Mask,
  stories,
  name: 'Highlighters/Mask',
  withRenderElementWrapper: true,
});

describe('mask highlighter', () => {
  it('should render custom highlighter', async () => {
    const { container } = render(
      <Mask padding={10} stroke={'red'} strokeWidth={5} strokeLinejoin={'bevel'}>
        <rect id="myRect" rx={10} ry={10} width={100} height={100} fill={'blue'} />
      </Mask>,
      { wrapper: simpleRenderElementWrapper }
    );

    await waitFor(() => {
      expect(container.querySelector('rect#myRect')).toBeInTheDocument();
      expect(container.querySelector('rect#myRect')?.getAttribute('fill')).toBe('blue');
    });
  });
});
