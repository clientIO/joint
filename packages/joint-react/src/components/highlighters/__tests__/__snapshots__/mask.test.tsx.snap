// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighters/Mask "Default": Highlighters/Mask-Default 1`] = `"<div style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-paper joint-theme-default" style="position: relative; width: 800px; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Highlighters/Mask "WithPadding": Highlighters/Mask-WithPadding 1`] = `"<div style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-paper joint-theme-default" style="position: relative; width: 800px; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Highlighters/Mask "WithSVGProps": Highlighters/Mask-WithSVGProps 1`] = `"<div style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-paper joint-theme-default" style="position: relative; width: 800px; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;
